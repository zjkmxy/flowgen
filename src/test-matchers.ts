import { execFileSync } from "child_process";

import chalk from "chalk";
import flow from "flow-bin";

import { beautify } from ".";

// eslint-disable-next-line @typescript-eslint/no-unused-vars,no-unused-vars
declare global {
  // eslint-disable-next-line @typescript-eslint/no-namespace
  namespace jest {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars,no-unused-vars
    interface Matchers<R> {
      toBeValidFlowTypeDeclarations(): R;
    }
  }
}

expect.extend({
  async toBeValidFlowTypeDeclarations(source) {
    const beautifiedSource = await beautify(source);
    try {
      execFileSync(
        flow,
        ["check-contents", "--color=always", "--timeout=30"],
        {
          input: beautifiedSource,
          stdio: ["pipe", "pipe", "pipe"],
        },
      );
    } catch (err) {
      return {
        message: () =>
          `expected ${chalk.bold(
            beautifiedSource.trimEnd(),
          )} to be valid flow:\n${chalk.red(err.stdout)}`,
        pass: false,
      };
    }

    return {
      message: () =>
        `expected ${chalk.bold(
          beautifiedSource.trimEnd(),
        )} not to be valid flow`,
      pass: true,
    };
  },
});
