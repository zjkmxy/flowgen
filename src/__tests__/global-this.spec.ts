import { compiler, beautify } from "..";
import "../test-matchers";

it("should not crash when getting globalThis in code", async () => {
  const ts = `import * as React from 'react';
export default class MenuStatefulContainer extends React.Component {
  handleItemClick: (
    event: React.MouseEvent<HTMLElement, globalThis.MouseEvent>
  ) => void = () => {};
  render(): React.ReactNode { return undefined; };
}
`;
  const result = compiler.compileDefinitionString(ts, { quiet: true });
  expect(await beautify(result)).toMatchSnapshot();
  expect(result).toBeValidFlowTypeDeclarations();
});
