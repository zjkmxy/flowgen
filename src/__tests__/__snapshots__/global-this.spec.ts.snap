// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should not crash when getting globalThis in code 1`] = `
"import * as React from "react";
declare export default class MenuStatefulContainer mixins React.Component<> {
  handleItemClick: (event: React.MouseEvent<HTMLElement, MouseEvent>) => void;
  render(): React.Node;
  undefined: /* NO PRINT IMPLEMENTED: SemicolonClassElement */ any;
}
"
`;
